atom/browser/api/event_subscriber.h

```diff
@@ -7,124 +7,50 @@
 
 #include <map>
 #include <string>
+#include <vector>
 
-#include "atom/common/api/event_emitter_caller.h"
-#include "base/synchronization/lock.h"
+#include "base/callback.h"
+#include "base/memory/weak_ptr.h"
 #include "content/public/browser/browser_thread.h"
-#include "native_mate/native_mate/arguments.h"
+#include "v8/include/v8.h"
+
+namespace atom {
+class URLRequestStreamJob;
+}
 
 namespace mate {
 
-namespace internal {
+class Arguments;
 
-class EventSubscriberBase {
+class EventSubscriber {
  public:
-  EventSubscriberBase(v8::Isolate* isolate, v8::Local<v8::Object> emitter);
-  virtual ~EventSubscriberBase();
-  virtual void EventEmitted(const std::string& event_name,
-                            mate::Arguments* args) = 0;
-
- protected:
-  void On(const std::string& event_name);
-  void Off(const std::string& event_name);
-  void RemoveAllListeners();
+  using EventCallback = base::Callback<void(mate::Arguments* args)>;
+
+  EventSubscriber(v8::Isolate* isolate,
+                  v8::Local<v8::Object> emitter,
+                  base::WeakPtr<atom::URLRequestStreamJob> url_job);
+  ~EventSubscriber();
 
  private:
-  std::map<std::string, v8::Global<v8::Value>>::iterator RemoveListener(
-      std::map<std::string, v8::Global<v8::Value>>::iterator it);
+  using JSHandlersMap = std::map<std::string, v8::Global<v8::Value>>;
 
-  v8::Isolate* isolate_;
-  v8::Global<v8::Object> emitter_;
-  std::map<std::string, v8::Global<v8::Value>> js_handlers_;
+  void On(const std::string& event, EventCallback&& callback);  // NOLINT
+  void Off(const std::string& event);
 
-  DISALLOW_COPY_AND_ASSIGN(EventSubscriberBase);
-};
+  void OnData(mate::Arguments* args);
+  void OnEnd(mate::Arguments* args);
+  void OnError(mate::Arguments* args);
```

Comment:
If we are going to make this class specific to Streams instead of generic event subscribers, maybe it should be renamed to `StreamSubscriber` or `StreamConsumer`?